node/core/pvf/src/error.rs

```diff
@@ -24,12 +24,50 @@ pub type PrepareResult = Result<Duration, PrepareError>;
 /// An error that occurred during the prepare part of the PVF pipeline.
 #[derive(Debug, Clone, Encode, Decode)]
 pub enum PrepareError {
+	/// An error that should trigger reliably. See [`DeterministicError`].
+	Deterministic(DeterministicError),
+	/// An error that may happen spuriously. See [`NonDeterministicError`].
+	NonDeterministic(NonDeterministicError),
+}
+
+impl fmt::Display for PrepareError {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		use PrepareError::*;
+		match self {
+			Deterministic(err) => write!(f, "{}", err),
+			NonDeterministic(err) => write!(f, "{}", err),
+		}
+	}
+}
+
+/// A deterministic error, i.e. one that should trigger reliably. Those errors depend on the PVF
+/// itself and the sc-executor/wasmtime logic.
+#[derive(Debug, Clone, Encode, Decode)]
+pub enum DeterministicError {
 	/// During the prevalidation stage of preparation an issue was found with the PVF.
 	Prevalidation(String),
 	/// Compilation failed for the given PVF.
 	Preparation(String),
 	/// An unexpected panic has occured in the preparation worker.
 	Panic(String),
+}
+
+impl fmt::Display for DeterministicError {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		use DeterministicError::*;
+		match self {
+			Prevalidation(err) => write!(f, "prevalidation: {}", err),
+			Preparation(err) => write!(f, "preparation: {}", err),
+			Panic(err) => write!(f, "panic: {}", err),
+		}
+	}
+}
+
+/// Non-deterministic errors can happen spuriously. Typically, they occur due to resource
+/// starvation, e.g. under heavy load or memory pressure. Those errors are typically transient but
+/// may persist e.g. if the node is run by overwhelmingly underpowered machine.
+#[derive(Debug, Clone, Encode, Decode)]
+pub enum NonDeterministicError {
```

Comment:
I get the naming of deterministic and non deterministic. However, I think this could be improved. I don't have a better terminology at hand right now, but I think it could be better.

Reply:
Yeah, it's awkward. Maybe `InternalPrepareError` and `InvalidPrepareError`? IDK.